hbase-server/src/main/java/org/apache/hadoop/hbase/master/region/MasterRegion.java

```diff
@@ -143,13 +148,23 @@ private void shutdownWAL() {
     }
   }
 
+  @RestrictedApi(explanation = "Should only be called in tests", link = "",
+      allowedOnPath = ".*/src/test/.*")
+  public void setTestFailure() {
+    this.updateFailForTest = true;
+  }
+
   public void update(UpdateMasterRegion action) throws IOException {
     try {
+      if (updateFailForTest) {
+        // test for HBASE-29251
+        throw new IOException("Update failed");
+      }
       action.update(region);
       flusherAndCompactor.onUpdate();
-    } catch (WALSyncTimeoutIOException e) {
-      LOG.error(HBaseMarkers.FATAL, "WAL sync timeout. Aborting server.");
-      server.abort("WAL sync timeout", e);
+    } catch (IOException e) {
+      LOG.error(HBaseMarkers.FATAL, "MasterRegion mutation is not successful. Aborting server.");
+      server.abort("MasterRegion mutation is not successful", e);
```

Comment:
Aborting is a start.
The rest of my question here is addressed by the discussion on the JIRA.